File: src/main/java/supercritical/api/unification/tag/TagPrefixExtension.java
package supercritical.api.unification.tag;

import java.util.function.Function;

/**
 * Extends GTCEu Modern's {@link com.gregtechceu.gtceu.api.data.tag.TagPrefix} to include
 * radiation damage behaviour for Supercritical nuclear items.
 */
public interface TagPrefixExtension {

    Function<Double, Double> getRadiationDamageFunction();

    void setRadiationDamageFunction(Function<Double, Double> function);

    static Function<Double, Double> getRadiationDamageFunction(com.gregtechceu.gtceu.api.data.tag.TagPrefix prefix) {
        return ((TagPrefixExtension) (Object) prefix).getRadiationDamageFunction();
    }

    static void setRadiationDamageFunction(com.gregtechceu.gtceu.api.data.tag.TagPrefix prefix,
                                           Function<Double, Double> function) {
        ((TagPrefixExtension) (Object) prefix).setRadiationDamageFunction(function);
    }
}

File: src/main/kotlin/supercritical/api/unification/tag/TagPrefixExtension.kt
package supercritical.api.unification.tag

import com.gregtechceu.gtceu.api.data.tag.TagPrefix

/**
 * Extends GTCEu Modern's [TagPrefix] to include
 * radiation damage behaviour for Supercritical nuclear items.
 */
typealias damageFunction = java.util.function.Function<Double, Double>

interface TagPrefixExtension {

    var radiationDamageFunction: damageFunction

    companion object {
        fun getRadiationDamageFunction(prefix: TagPrefix): damageFunction {
            return (prefix as Any? as TagPrefixExtension).radiationDamageFunction
        }

        fun setRadiationDamageFunction(
            prefix: TagPrefix,
            function: damageFunction
        ) {
            (prefix as Any? as TagPrefixExtension).radiationDamageFunction = function
        }
    }
}
